content/blog/sivayanama-testing-opsramp-apis-with-pytest-fixtures.md

```diff
@@ -0,0 +1,242 @@
+
+```python
+import pytest
+
+@pytest.fixture
+def hello_world():
+    return 'Hello, Happy testing'
+```
+
+
+
+## Invoke fixtures
+
+Test functions can invoke fixtures just by declaring the required fixtures as arguments. 
+
+```python
+def test_hello_world(hello_world):
+    assert hello_world == 'Hello, Happy testing'
+```
+
+## Invoke OpsRamp API
+
+Let us try to execute OpsRamp [Get Access Token API](https://develop.opsramp.com/v2/api/auth/tenancy-auth-oauth-token)  as shown below.
+
+```python
+import http.client
+import json
+import traceback
+import urllib
+from asyncio.log import logger
+
+import pytest
+
+def get_auth_header(api_endpoint, client_id, client_secret, grant_type="client_credentials"):
+    """
+    Returns bearer token string in the blow format
+    bearer adfa-afdaf-1599-402c-a3ee-1ed24f597cc8
+    """
+
+    api_connection = http.client.HTTPSConnection(api_endpoint)
+    params = {"client_id": client_id, "client_secret": client_secret, "grant_type": "client_credentials"}
+
+    payload = urllib.parse.urlencode(params)
+
+    headers = {
+        'accept': "application/json",
+        'content-type': "application/x-www-form-urlencoded",
+    }
+
+```
+
+## Define Fixture to get Access Token
+
+You can transform above function as PyTest Fixture by simply decorating the code as shown below 
+
+
+
+@pytest.fixture
+
+def get_auth_header(api_endpoint, client_id, client_secret, grant_type="client_credentials"):
+
+
+
+## Autouse fixtures
+
+Sometimes some of your fixtures are required for all other tests. The above Get acces token is perfect use case for this. Every API call is depending on this API. In this case you can make this fixture as “autouse fixture”  by passing in autouse=True to the fixture’s decorator. The autouse fixure capability will reduce lot of reduanct requests. 
+
+```python
+@pytest.fixture(autouse=True)
+def get_auth_header(api_endpoint, client_id, client_secret, grant_type="client_credentials"):
+```
+
+
+
+## Provisioing of third-party plugin Fixtures 
+
+Not necessarily you have to define all fixtures on your own code. Fixtures can be provided by third party plugins as well and you are free to use them. You just need to install the required plugins. We will see an example with  pytest-datadir plugin
+
+
+
+Install this plugin as shown below 
+
+```shell
+pip install pytest-datadir
+```
+
+
+
+```python
+@pytest.fixture(autouse=True)
+
+def get_bearer_token(shared_datadir,request):
+
+json_path = (shared_datadir / "auth_header_input.json")
+```
+
+![](https://lh4.googleusercontent.com/dacTgDdw17BzeyCitShA73WSip9LVtenQoNN-uraaN5tKEU5cA_xP3cEmNPWmTzU3A1HegdoOVvwPbyYqQuoLeEk4W766nIvpBdoTzUdIiT2dXiOQG0_h7atQWS7-T9qvRrieuhlEV84VS15ir11Ocw)
+
+
+
+get_bearer_token  fixture is using [shared_datadir](https://pypi.org/project/pytest-datadir/) fixture from third party plugin. Shared_datadir fixture provision  data folder from the current folder as pathlib.Path object. 
```

Comment:
Use escape character (a backslack) just before the "underscore" for the function "get_bearer_token".
There is an extra space between the get_header_token and fixture.
Extra space character between provision and data folder.